<!DOCTYPE HTML>
<p>The GTagHD Web Tool can automatically generate short homology arms for high efficiency, CRISPR-mediated, precise integration of pGTag and pPRISM vectors to your loci of interest. </p>
<p></p>
<p>To use the tool, choose the "Submit Single Job" tab.</p>
<br>
<p>Step 1) Choose to use either our universal guide RNA for liberation of the integration cassette (recommended), or submit a custom guide RNA for cassette liberation (this can be the same as the guide for your loci.)</p>
<br>
<p>Step 2) Enter your CRISPR target sequence (including the PAM). This sequence should be 20 bp long. DO NOT include the PAM sequence. GTagHD assumes that your nuclease behaves similarly to Cas9 proteins. Thus, your nuclease target will generate a double-strand break between the 17th and 18th nucleotides of your 20 bp target sequence if you choose the 'sense' option:</p>
<div class = "image1">
  <p><img src="senseExample.png" alt = "sense example"></p>
  <p class="caption">
    <i>Example of an SpCas9 target sequence (blue) and PAM sequence (red) in the 5'-3' orientation. GTagHD requires only the blue target sequence. GTagHD is therefore PAM agnostic.</i></p>
</div>
<br>
<p>And the double-strand break will be between the 3rd and 4th nucleotides if you choose the 'anti-sense' option:</p>
<div class = "image1">
  <p><img src = "antisenseExample.png" alt = "antisense example"></p>
  <p class="caption">
    <i>Example of an SaCas9 PAM sequence (red) and target sequence (blue) in the 5'-3' orientation. This target sequence will induce a double-strand break on the 3'-5' (antisense) strand.</i></p>
</div>
<br>
<p>Note: guide sequences generated by other webtools may have 1-2 bases at the 5' end of the guide that are changed to Gs for T7 <i>in vitro</i> purposes. Please input the target sequence as it appears in the genome.</p>
<br>
<p>Step 3) Enter the complete cDNA containing your CRISPR target from Step 2. Please note that your CRISPR target should appear exactly once in either the forward or reverse complement direction of the cDNA sequence.</p>
<p>*We will soon support automatically downloading a cDNA sequence based on ENSEMBL gene ID.</p>
<br>
<p>Step 4) Specify the number of 'padding' nucleotides (0-2) to be added to the targeting oligos such that the resulting integration will be in frame.</p>
<br>
<p>Step 5) Select the length of homology to be used. We recommend using the default value of 48 nucleotides.</p>
<p>If there are any problems with the sequences and values you have entered, the web page will inform you of the errors and give you advice on how to fix them.</p>
<br>
<p>If no errors are present, you can press the 'Submit' button at the bottom of the page to generate your targeting oligos.</p>
<br>
<p>You can also see an example of the inputs using the 'Example' link on the left hand column of the 'Submit Single Job' page.</p>
<p>For further information, see <a href = "Designing Homology Oligos for Universal Vectors_JMW_(060117).pdf" target="_blank"> Designing Homology Oligos for Universal Vectors</a></p>
<br>
