<!DOCTYPE HTML>
<html>
<body>
  <h3 style = "display:inline-block">Instructions  </h3><a href="#instructions" data-toggle = "collapse"> expand/collapse </a>
<div id = "instructions" class = "collapse in">
<p>The <b>GTagHD</b> web tool can automatically generate short homology arms for high efficiency, CRISPR-mediated, precise integration of pGTag vectors and pPRISM vectors <i>(coming soon)</i> to your loci of interest. </p>
<br>
<p>To use the tool, choose the "Submit Single Job" tab.</p>
<br>
<p>Step 0) Choose which tool/plasmid series you are interested in using. The Iowa State/Mayo Clinic Alliance for Genome Engineering is actively developing several plasmid series for precision genome editing. We currently support two plasmid series, pGTag and pPRISM. More plasmid series will be added in the future.</p>
<br>
<p>Step 1) Choose to use either our universal guide RNA for liberation of the integration cassette (recommended), or submit a custom guide RNA for cassette liberation (this can be the same as the guide for your loci.)</p>
<br>
<p>Step 2) Enter your CRISPR target sequence. This sequence should be 20 bp long. DO NOT include the PAM sequence. GTagHD assumes that your nuclease behaves similarly to Cas9 proteins. Thus, your nuclease target will generate a double-strand break between the 17th and 18th nucleotides of your 20 bp target sequence if you choose the 'sense' option:</p>
<div class = "image1">
  <p><img src="senseExample.png" alt = "sense example"></p>
  <p class="caption">
    <i>Example of an SpCas9 target sequence (blue) and PAM sequence (red) in the 5'-3' orientation. GTagHD requires only the blue target sequence. GTagHD is therefore PAM agnostic.</i></p>
</div>
<br>
<p>And the double-strand break will be between the 3rd and 4th nucleotides if you choose the 'anti-sense' option:</p>
<div class = "image1">
  <p><img src = "antisenseExample.png" alt = "antisense example"></p>
  <p class="caption">
    <i>Example of an SpCas9 PAM sequence (red) and target sequence (blue) in the 5'-3' orientation. This target sequence will induce a double-strand break on the 3'-5' (antisense) strand.</i></p>
</div>
<br>
<p>Note: guide sequences generated by other webtools may have 1-2 bases at the 5' end of the guide that are changed to Gs for T7 <i>in vitro</i> purposes. Please input the target sequence as it appears in the genome.</p>
<br>
<p>Step 3) Enter the complete DNA sequence (including introns/non-coding regions) containing your CRISPR target from Step 2. You can do this by entering a RefSeq or GenBank* accession corresponding to your DNA sequence of interest OR by copying and pasting a plain-text sequence (no FASTA headers, please).</p>
<p>Please note that your CRISPR target should appear <i>exactly once</i> in either the 5' or 3' strand of the sequence.</p>
<p>If you are copying/pasting a sequence, you must provide enough sequence context surrounding the target to generate homology arms; for instance, if you want to use 48bp of homology, you should provide at least 50bp of context upstream and downstream of the double-strand break site in your DNA sequence.</p>
<p>*We will soon support automatically downloading a DNA/gene sequence based on Ensembl gene ID.</p>
<br>
<p>Step 3a) Specify the number of additional nucleotides (0-2) to be added to the targeting oligos such that the resulting integration will be in frame.</p>
<div class = "image2">
  <p><img src = "codonBreakExample.png" alt = "codon break repair example"></p>
  <p class = "caption">
    <i>In this example SpCas9 target sequence, the double-strand break site is between the second and third nucleotides in the seventh codon. Thus, targeting this site would induce a frameshift in the integration site. To prevent this, you should select a single additional nucleotide. If the break was between the first and second nucleotide of the codon, you would specify two additional nucleotides.</i>
  </p>
</div>
<br>
<p>Step 4) Select the length of homology to be used. We recommend using the default value of 48 nucleotides for best results. Please keep in mind the sequence length requirements mentioned in Step 3.</p>
<br>
<p>Step 5) In this box, you can specify the pGTag plasmid you are utilizing and generate a GenBank format file containing the plasmid sequence with your incorporated oligonucleotides. We will shortly be adding pPRISM plasmid here, as well. To use this feature, hit the "Download Plasmid File" button which will appear below your oligonucleotides after you have hit the "Submit" button for the first time. You can download a different plasmid <b>WITHOUT</b> hitting the "Submit" button again by choosing a different plasmid from the drop-down menu, and then hitting "Download Plasmid File" again.</p>
<br>
<p>If there are any problems with the sequences and values you have entered, the web page will inform you of the errors and give you advice on how to fix them.</p>
<br>
<p>If no errors are present, you can press the 'Submit' button at the bottom of the page to generate your targeting oligos. Your oligonucleotides will appear very quickly. If you are using a GenBank/RefSeq accession, it may take a few seconds to retrieve your accession.</p>
<br>
<p>You can also see an example of the inputs using either of the 'Example' links on the left hand column of the 'Submit Single Job' page.</p>
<p>For further information, see <a href = "Designing Homology Oligos for Universal Vectors_JMW_(060117).pdf" target="_blank"> Designing Homology Oligos for Universal Vectors</a></p>
</div>
<br>
<h3 style = "display:inline-block">Frequently Asked Questions (FAQs)  </h3><a href="#FAQs" data-toggle = "collapse"> expand/collapse </a>
<div id = "FAQs" class = "collapse in">
  <p><b>Does GTagHD store data inputs?</b></p>
  <p>We do not store any data inputs. Your data inputs are only available to GTagHD for as long as your GTagHD session is active, and are only used to generate your results during that time. The file containing your oligonucleotides is only generated when you click the "Download Oligos" button. The file containing the oligos incorporated into a plasmid is also dynamically generated when you click "Download Plasmid File." These files are not stored on our server. Once you exit the GTagHD browser window, your session is deleted and everything you did during the session is gone. Therefeore, if you do not download or otherwise save your results, you will have to perform your analysis again.</p>
  <p><b>I chose the option to automatically generate nucleotides to repair a codon break, but no additional nucleotides were generated. Why did this happen?</b></p>
  <p>This can occur if:</p>
  <li>Your GenBank/RefSeq accession is missing exon information</li>
  <li>Your GenBank/RefSeq accession has improperly annotated exon information</li>
  <li>Your guide RNA corresponds to an intron (and thus there is no codon break to repair</li>
  <p></p>
  <p>We will shortly have an option to manually specify the number of nucleotides to generate. You can also manually paste your sequence of interest, and manually generate codon break repair nucleotides that way.</p>
</div>
<br>
<h3 style = "display:inline-block">Troubleshooting  </h3><a href="#troubleshooting" data-toggle = "collapse"> expand/collapse </a>
<div id = "troubleshooting" class = "collapse in">
  <p>If you experience trouble or the website crashes, please submit a bug report using the "Report Bugs or Contact Us" tab, in the upper-left portion of the navigation bar.</p>
</div>
</body>
</html>
