<!DOCTYPE HTML>
<html>
<body>
  <h3 style = "display:inline-block">Instructions  </h3><a href="#instructions" data-toggle = "collapse"> expand/collapse </a>
<div id = "instructions" class = "collapse in">
<p>The <b>GTagHD</b> web tool can automatically generate short homology arms for high-efficiency, CRISPR-mediated, precise integration of pGTag vectors and pPRISM vectors to your loci of interest. <b>GTagHD</b> should be used in conjunction with the <b>GeneWeld</b> protocols outlined by <a href="https://doi.org/10.1101/431627" target="_blank">Wierson and Welker et al.</a></p>
<br />
<p>The <b>About</b> tab gives a brief overview of how the <b>GeneWeld</b> technology works.</p>
<p>The <b>Instructions and FAQs</b> tab contains detailed instructions and additional information for each step of running GTagHD. Please see the <a href="https://github.com/Dobbs-Lab/GTagHD/blob/master/README.md" target="_blank">GitHub README</a> for instructions on how to install and run GTagHD locally.</p>
<p>To use the tool, choose the <b>Submit Job</b> tab.</p>
<br />
<p><b>Step 0) Choose which GeneWeld tool/plasmid series you are using.</b> The Iowa State/Mayo Clinic Alliance for Genome Engineering is actively developing several plasmid series for precision genome editing. We currently have two plasmid series available - the pGTag series and the pPRISM series. The pGTag series plasmids are available via <a href="https://www.addgene.org/Jeffrey_Essner/" target="_blank">AddGene</a>; pPRISM plasmids will be added soon.</p>
<p>The drop-down menu for this step has three options: (1) pGTag, (2) pPRISM, and (3) Custom. The pGTag and pPRISM vector series use different restriction enzymes in the cloning step, and thus differ primarily in the overhangs generated in the processing phase. The "Custom" option is for use with non-GeneWeld plasmids (not recommended), and does not automatically add restriction enzyme site overhangs to the oligonucleotide products.</p>

<br />
<p><b>Step 1) Choose the <i>plasmid</i> gRNA</b> - that is, the CRISPR guide RNA used to liberate the integration cassette from the plasmid. This step has two radio buttons: the "Universal Guide RNA" and "Custom Guide RNA" options. All GeneWeld plasmids use the "Universal Guide RNA" - a sequence without a homologous site in any known vertebrate genome - to liberate plasmid cargos. Since the universal gRNA is unique to the GeneWeld plasmid, this reduces the possibility of the plasmid gRNA producing off-target effects in the genome.</p> 
<p>If you choose to use a custom plasmid (again, not recommended), choose the "Custom Guide RNA" radio button, which will allow you to input your plasmid gRNA in a text box. This information is used to add your desired gRNA to the oligonucleotides for incorporation into the plasmid. Please use IUPAC standard nucleotides (A, C, G, T); ambiguous characters are not allowed. <strong>Use the "Universal Guide RNA" option if your plasmid gRNA is ALREADY present in your chosen plasmid</strong>, and you therefore DO NOT need to clone it into your plasmid.</p>
<br />
<p><b>Step 2) Enter your <i>genomic</i> gRNA.</b> This sequence should be the "base" 20 bp long gRNA sequence, and <b>exactly match</b> your genomic sequence (e.g., some guide sequences generated by other webtools may have 1-2 bases at the 5' end of the guide that are changed to 'G's for T7 promoters. Please input the target sequence as it appears in the genome.) GTagHD uses this information to figure out where the double-strand break will occur in the genome, so that it can extract the homologous sequence. DO NOT include the PAM sequence. Instead, use the radio buttons below the text box to indicate whether your PAM is upstream or downstream of your gRNA - e.g., if you are using CRISPR/Cas9, would the PAM sequence appear as "CCN-gRNA" or as "gRNA-NGG". This gives GTagHD the necessary information to figure out the orientation of your gRNA so that the integration is oriented correctly.</p> 
<p>GTagHD assumes that your nuclease behaves similarly to Cas9 proteins. Thus, your nuclease target will generate a double-strand break between the 17th and 18th nucleotides of your 20 bp target sequence if you choose the '3' downstream' option:</p>
<div class = "image1">
  <p><img src="senseExample.png" alt = "sense example"></p>
  <p class="caption">
    <i>Example of an SpCas9 target sequence (blue) and PAM sequence (red) in the 5'-3' orientation. GTagHD requires only the blue target sequence. GTagHD is therefore PAM agnostic.</i></p>
</div>
<br />
<p>And the double-strand break will be between the 3rd and 4th nucleotides if you choose the '5' upstream' option:</p>
<div class = "image1">
  <p><img src = "antisenseExample.png" alt = "antisense example"></p>
  <p class="caption">
    <i>Example of an SpCas9 PAM sequence (red) and target sequence (blue) in the 5'-3' orientation. This target sequence will induce a double-strand break on the 3'-5' (antisense) strand.</i></p>
</div>
<br />
<p><b>Step 3) Enter the complete <i>genomic</i> DNA sequence (including introns/non-coding regions) containing your gRNA from Step 2.</b> Choose a radio button to use a GenBank or RefSeq accession, and Ensembl accession, or copy and paste a sequence.</p>
<p>For GenBank and RefSeq accessions, please specify a nucleotide sequence. Gene IDs may have many transcripts associated with them, and GTagHD can't generate precise oligonucleotides without a specific seqence. We also <strong>strongly</strong> recommend that you do NOT use mRNA transcript IDs - if the required homology length spans across a splice site, this will make the oligonucleotide homology span the splice site as well - resulting in non-homologous nucleotides and a failed knock-in.</p>
<p>For Ensembl accessions, please specify a transcript, exon, or protein ID. As with GenBank and RefSeq accessions, Ensembl gene IDs may have many associated transcripts, while transcript, exon, and protein IDs only have a single associated sequence.</p>
<p>When copy/pasting a sequence, please paste in plain text and do not include FASTA headers. Sequence must use IUPAC standard nucleotides and cannot contain ambiguous characters. You must also provide enough sequence context surrounding the target to generate homology arms; for instance, if you want to use 48bp of homology, you should provide at least 50bp of context upstream and downstream of the double-strand break site in your DNA sequence.</p>
<p>Please note that your genomic gRNA should appear <i>exactly once</i> in either the 5' or 3' strand of the sequence, regardless of input choice. GTagHD <b>does not</b> automatically check for off-target sites. There are other tools available for this purpose, like <a href="http://www.crisprscan.org/" target="_blank">CRISPRScan</a>.</p>
<br />
<p><b>Step 3a) Specify nucleotides to repair a codon break in the genomic locus.</b> If you use a GenBank, RefSeq, or Ensembl accession, GTagHD can automatically determine if your genomic gRNA target induces a DSB in a codon such that a frameshift would be introduced. You can choose to automatically repair this break (recommended.) If this option is selected, GTagHD will add nucleotides to your homology oligonucleotides such that the genomic target will remain in frame, with the codon repaired to create the original amino acid.</p>
<p>If you use a pasted sequence, you can manually select the number of additional nucleotides required to maintain the reading frame (0 - the DSB is between codons; 1 - the DSB occurs between the second and third nucleotides in the codon; 2 - the DSB occurs between the first and second nucleotides in the codon.</p>
<div class = "image2">
  <p><img src = "codonBreakExample.png" alt = "codon break repair example"></p>
  <p class = "caption">
    <i>In this example SpCas9 target sequence, the double-strand break site is between the second and third nucleotides in the sixth codon displayed. Thus, targeting this site would induce a frameshift upon integration. To prevent this, you should select a single additional nucleotide, and GTagHD will add the appropriate nucleotide(s) (in this case, a single adenine) to the oligonucleotide sequences. If the break was between the first and second nucleotide of the codon, you would specify two additional nucleotides.</i>
  </p>
</div>
<br />
<p><b>Step 4) Select the length of homology to be used.</b> This will be the length of the homology between each end of the plasmid and the genomic locus. We recommend using the default value of 48 nucleotides for best results. Please keep in mind the sequence length requirements mentioned in Step 3 - if your genomic DNA provided in Step 3 doesn't have context, an error will occur.</p>
<br />
<p><b>Step 5) (Optional) Specify the pGTag plasmid you are utilizing and generate a GenBank format file containing the plasmid sequence with your incorporated oligonucleotides.</b> <!--We will shortly be adding pPRISM plasmid here, as well.--> This step has no affect whatsoever on your oligonucleotides. This just gives you the ability to download a GeneWeld plasmid map with your (labeled) oligonucleotides incorporated into the map. To use this feature, hit the "Download Plasmid File" button which will appear below your oligonucleotides after you have hit the "Submit" button for the first time. You can download a different plasmid <b>WITHOUT</b> hitting the "Submit" button again by choosing a different plasmid from the drop-down menu, and then hitting "Download Plasmid File" again. We will be adding additional GeneWeld plasmids to this menu as they become available.</p>
<p>The plasmid maps are in A Plasmid Editor (ApE) format, which is completely compatible with GenBank format (gb) and any tools that use GenBank format.</p>
<br />
<p>If there are any problems with the sequences and values you have entered, the GTagHD will attempt to identify the problem and give you advice on how to fix them.</p>
<br />
<p>If no errors are present, you can press the 'Submit' button at the bottom of the page to generate your targeting oligos. Your oligonucleotides will appear very quickly. If you are using a GenBank/RefSeq/Ensembl accession, it may take a few seconds to retrieve the corresponding sequence.</p>
<p>If the screen grays out at this stage, congratulations! You have found an error we didn't think of and haven't accounted for yet, and we would GREATLY appreciated it if you filled out a bug report form <a href="http://genesculpt.org/bugreports.html" target="_blank">here</a> to tell us about it. You can also refresh this page and then submit a report using the "Report Bugs or Contact Us" tab.</p>
<br />
<p>Once GTagHD finishes processing the information, it will output four sequences: 5' forward, 5' reverse, 3' forward, and 3' reverse oligonucleotide sequences. The restriction enzyme sites for cloning into your plasmid will be lowercase, while the rest of the sequence will be UPPERCASE. These oligonucleotides are ready-to-order, and can be downloaded in a text file after hitting the "Download Oligos" button.</p>
<p>Please see the <a href="2019_Wierson_et_al_GeneWeld_Protocol.pdf" target="_blank">GeneWeld protocol</a> for detailed instructions on the GeneWeld protocol.</p>
<br />
<p>You can also see an example of the inputs using any of the 'Example' links on the left hand column of the 'Submit Single Job' page. The GenBank and Ensembl example IDs belong to the zebrafish "Floating Head" (flh) or <i>noto</i> transcript.</p>
<p>For further information on the algorithm for generating the oligonucleotides, see <a href = "Designing Homology Oligos for Universal Vectors_JMW_(011119).pdf" target="_blank"> Designing Homology Oligos for Universal Vectors</a>.</p>
</div>
<br />
<h3 style = "display:inline-block">Frequently Asked Questions (FAQs)  </h3><a href="#FAQs" data-toggle = "collapse"> expand/collapse </a>
<div id = "FAQs" class = "collapse in">
  <!--
  <p><b>The site isn't quite working properly; how can I fix this?</b></p>
  <p>GTagHD has been tested on the following browsers: </p>
  
  
  <p>If you are not using one of these browsers, we highly recommend you do so and see if that resolves the problem.</p>-->
  <p><b>Does GTagHD store data inputs?</b></p>
  <p>We do not store any data inputs. When you visit GTagHD, you are interacting with a Docker container created from the Docker GTagHD image. This container only exists while your session is open (i.e., while your web browser is pointed at GTagHD.) Thus, your data inputs are only available to GTagHD for as long as your GTagHD session is active, and are only used to generate your results during that time. The file containing your oligonucleotide sequences is only generated when you click the "Download Oligos" button. The file containing the oligos incorporated into a plasmid is also dynamically generated when you click "Download Plasmid File." Both of these files are created within the temorary Docker container. Once you exit the GTagHD browser window, your session (Docker container) is deleted 60 seconds later, and since everything you did during the session is within that container, it is deleted as well. Therefeore, if you do not download or otherwise save your results, you will have to perform your analysis again.</p>
  <br />
  <p><b>I chose the option to automatically generate nucleotides to repair a codon break, but no additional nucleotides were generated. Why did this happen?</b></p>
  <p>This can occur if:</p>
  <li>Your GenBank/RefSeq/Ensembl accession is missing exon information</li>
  <li>Your GenBank/RefSeq/Ensembl accession has improperly annotated exon information</li>
  <li>Your guide RNA corresponds to an intron (and thus there is no codon break to repair)</li>
  <p></p>
  <p>If this situation occurs, you can manually paste your sequence of interest, and manually generate codon break repair nucleotides that way.</p>
  <br />
  <p><strong>Why shouldn't I use RefSeq mRNA entries?</strong></p>
  <p>RefSeq mRNA entries usually contain the post-spliced DNA sequence. GTagHD creates oligonucleotides based on the sequence it is given, so as far as it can tell, the post-spliced RefSeq mRNA has no introns. If your guide corresponds to a position in the exon where the homology arms would run off the end of the exon (e.g., your genomic guide RNA cuts between the 28th and 29th nucleotides in the exon, and you select 48 bp of homology), GTagHD pulls nucleotides from the adjacent exon, rather than the adjacent intron. <strong>TL;DR: if you use RefSeq mRNA accessions, this may result in the generation of oligonucleotides that are not homologous to your genome, and thus will not generate precise integrations.</strong> We are discussing ways to address this issue in future GTagHD releases.</p>
</div>
<br />
<h3 style = "display:inline-block">Troubleshooting  </h3><a href="#troubleshooting" data-toggle = "collapse"> expand/collapse </a>
<div id = "troubleshooting" class = "collapse in">
  <p><strong>If you experience trouble or the website crashes, please submit a bug report using the "Report Bugs or Contact Us" tab, in the upper-right portion of the navigation bar.</strong></p>
</div>
</body>
</html>
